libvmi/driver/xen/xen_events.c

```diff
@@ -1104,6 +1129,93 @@ status_t process_desc_access(vmi_instance_t vmi, vm_event_compat_t *vmec)
     return VMI_SUCCESS;
 }
 
+static int check_domain_shutdown(
+    void *key,
+    void *value,
+    void *data)
+{
+    uint32_t *pd = key;
+    xen_check_domain_t *in = data;
+
+    if ( !key || !value || !data )
+        return 1;
+
+    if (!in->xen.libxsw.xs_is_domain_introduced(in->xen.xshandle, *pd)) {
+        in->domain = *pd;
+        in->found = true;
+        in->uuid = value;
+    }
+    return 1;
+}
+
+static
+status_t process_domain_watch(vmi_instance_t vmi, vm_event_compat_t* UNUSED(vmec))
+{
+    xen_instance_t *xen = xen_get_instance(vmi);
+    unsigned int num;
+    char **vec = xen->libxsw.xs_read_watch(xen->xshandle, &num);
+    status_t ret = VMI_FAILURE;
+
+    if (vec && !strncmp(vec[XS_WATCH_TOKEN], INTRODUCE_TOKEN, sizeof(INTRODUCE_TOKEN))) {
+
+        int domid = 1, err = -1;
+        xc_dominfo_t dominfo;
+
+        while (( err = xen->libxcw.xc_domain_getinfo(xen->xchandle, domid, 1, &dominfo)) == 1) {
+            domid = dominfo.domid + 1;
+            if (xen->libxsw.xs_is_domain_introduced(xen->xshandle, dominfo.domid)) {
+                void *uuid = g_tree_lookup(xen->domains, &dominfo.domid);
+
+                if (!uuid) {
+                    //get uuid for the new domain
+                    char path[512];
+                    char *tmp;
+                    unsigned int size = 0;
+
+                    snprintf(path, sizeof( path ), "/local/domain/%d/vm", dominfo.domid);
+                    tmp = xen->libxsw.xs_read(xen->xshandle, XBT_NULL, path, &size);
+                    if (tmp || (sizeof(tmp) > 4)) {
+                        uint32_t *domid = malloc(sizeof(uint32_t));
+                        char *uuid = strdup(tmp + 4);
+
+                        *domid = dominfo.domid;
+                        g_tree_insert(xen->domains, domid, uuid);
```

Comment:
Could you explain why we need that tree to store the domain's in? Why not just pass the event straight through to the registered event handler?

Reply:
Never mind, I see it's just to look up the uuid from instead of doing a read from Xenstore. I don't think that's necessary. Is that lookup going to be that much faster then doing the read from xenstore?

Reply:
It's probably faster, but that's not the main reason. The problem is that on `@introduceDomain` we can just call `xs_is_domain_introduced()` and figure out which domains are new. But on `@releaseDomain`, the domain is gone, so iterating over all the available domains to find the one that has finished is pointless (only the active domains will show up).

Reply:
Gotcha, makes sense.